pypeit/flatfield.py

```diff
@@ -382,7 +318,552 @@ def load(self, ifile=None):
+
+            # Create the tilts image for this slit
+            tilts = tracewave.fit2tilts(rawflat.shape, self.tilts_dict['coeffs'][:,:,slit],
+                                        self.tilts_dict['func2d'])
+            # Convert the tilt image to an image with the spectral pixel index
+            spec_coo = tilts * (nspec-1)
+
+            # Only include the trimmed set of pixels in the flat-field
+            # fit along the spectral direction.
+            spec_gpm = onslit_trimmed & gpm_log # & (rawflat < nonlinear_counts)
+            spec_nfit = np.sum(spec_gpm)
+            spec_ntot = np.sum(onslit)
+            msgs.info('Spectral fit of flatfield for {0}/{1} '.format(spec_nfit, spec_ntot)
+                      + ' pixels in the slit.')
+            # Set this to a parameter?
+            if spec_nfit/spec_ntot < 0.5:
+                # TODO: Shouldn't this raise an exception or continue to the next slit instead?
+                msgs.warn('Spectral fit includes only {:.1f}'.format(100*spec_nfit/spec_ntot)
+                          + '% of the pixels on this slit.' + msgs.newline()
+                          + '          Either the slit has many bad pixels or the number of '
+                            'trimmed pixels is too large.')
+
+            # Sort the pixels by their spectral coordinate.
+            # TODO: Include ivar and sorted gpm in outputs?
+            spec_gpm, spec_srt, spec_coo_data, spec_flat_data \
+                    = flat.sorted_flat_data(flat_log, spec_coo, gpm=spec_gpm)
+            # NOTE: By default np.argsort sorts the data over the last
+            # axis. Just to avoid the possibility (however unlikely) of
+            # spec_coo[spec_gpm] returning an array, all the arrays are
+            # explicitly flattened.
+            spec_ivar_data = ivar_log[spec_gpm].ravel()[spec_srt]
+            spec_gpm_data = gpm_log[spec_gpm].ravel()[spec_srt]
+
+            # Rejection threshold for spectral fit in log(image)
+            # TODO: Make this a parameter?
+            logrej = 0.5
+
+            # Fit the spectral direction of the blaze.
+            # TODO: Figure out how to deal with the fits going crazy at
+            # the edges of the chip in spec direction
+            # TODO: Can we add defaults to bspline_profile so that we
+            # don't have to instantiate invvar and profile_basis
+            spec_bspl, spec_gpm_fit, spec_flat_fit, _, exit_status \
+                    = utils.bspline_profile(spec_coo_data, spec_flat_data, spec_ivar_data,
+                                            np.ones_like(spec_coo_data), inmask=spec_gpm_data,
+                                            nord=4, upper=logrej, lower=logrej,
+                                            kwargs_bspline={'bkspace': spec_samp_fine},
+                                            kwargs_reject={'groupbadpix': True, 'maxrej': 5})
+
+            if exit_status > 1:
```

Comment:
Okay, did a few things:
 - If this bspline fit fails, the code simply continues on to the next slit and warns the user that the flat-field correction isn't applied to this slit.  By continuing, the flat arrays just have 1s in the region of this slit (see the instantiation before the loop over slits).
 - If the bspline fit to the illumination profile fails, the code warns the user and leaves `msillumflat` as is, meaning it will have 1s for this slit, and continues on to the  2D fit.
 - If the 2D fit fails, the code warns the user and leaves the 2D model as 1s and proceeds to set `mspixelflat` based on whatever fits were successful.

Happy to change any of these if you want a different behavior.

Reply:
That sounds right to me.